### ISSUE TITLE
Support WORDPRESS_CONFIG_EXTRA for custom env variables

### ISSUE THREAD

**Issue by lwj5** [bitnami/bitnami-docker-wordpress#167]
**Description**

For extra Wordpress configuration variables that are currently not explicitly defined, create a environment variable that takes in a list of key:value that would be loaded into wp-config.php.

See official wordpress docker PR here:
https://github.com/docker-library/wordpress/pull/142

**Describe the results you expected:**

Expect these variables to load into wp-config.php.

**Comment by juan131** [comment]
Hi @lwj5 

Thanks so much for this suggestion.

In the past, I already mentioned this possibility on this issue: https://github.com/bitnami/bitnami-docker-wordpress/issues/131

> As I see there are two ways to make it easier:
> 
> A) We implement a way to append custom configuration to wp-config.php.
> B) We implement a specific env. variable to configure the WP_MEMORY_LIMIT setting.

@javsalgar @tompizmor don't you think we should implement this on the WP image? I'd provide the user a lot flexibility to customize the **wp-config.php**.

**Comment by lwj5** [comment]
Hi @juan131

It’s good to hear you are exploring different options.

I’m running on K8s with persistent storage and I’m wondering if there’s any chance the wp-config.php will be replaced if I edit it manually?

**Comment by juan131** [comment]
Hi @lwj5 

The **wp-config.php** is located under the directory `/bitnami/wordpress/`. If you're using the WordPress Helm Chart (stable/wordpress) and you're enabled the persistence, then the configuration file is persisted.

Therefore, if you edit the file manually, those changes will be persisted.

**Comment by javsalgar** [comment]
I agree that we should add as much flexibility as possible to the WordPress image. Option A) would fit better IMHO

**Comment by polRk** [comment]
```
foreach ($_ENV as $key => $value) {
	$capitalized = strtoupper($key);
	if (!defined($capitalized)) {
		define($capitalized, $value);
	}
}
```

**Comment by juan131** [comment]
Thanks for the tip @polRk !! It can be a good trip to support every config setting as env var.
The approach we were originally thinking about is even more flexible since it'll allow you to append a piece of configuration to the `wp-config.php` file

**Comment by leandropapasidero** [comment]
Is `WORDPRESS_EXTRA_WP_CONFIG_CONTENT` supported? I am setting the env variable but it doesn't change `wp-content` at all.

docker-compose.yaml
```
...
      - WORDPRESS_EXTRA_WP_CONFIG_CONTENT= |
        /* Multisite */
        define('WP_ALLOW_MULTISITE', true );
        define('MULTISITE', true);
        define('SUBDOMAIN_INSTALL', false);
        define('DOMAIN_CURRENT_SITE', '127.0.0.1');
        define('PATH_CURRENT_SITE', '/');
        define('SITE_ID_CURRENT_SITE', 1);
        define('BLOG_ID_CURRENT_SITE', 1);
...
```

I can see the env variable value from inside of the container if I run `env`

**Comment by juan131** [comment]
Hi @leandropapasidero 

I was not able to reproduce the issue. I added the same env. variable with the same piece of code and then I run the commands below:

```console
$ docker-compose up -d 
$ docker-compose exec wordpress cat /opt/bitnami/wordpress/wp-config.php
```

The content was properly added to the `wp-config.php` file.

Did you add the environment variable on a clean environment? I mean, without existing data on the volumes attached to WordPress.

**Comment by leandropapasidero** [comment]
@juan131 Thank you so much for your reply. I just noticed that it will only work on a new volume (first time setup). Once you run the container, it won't change `wp-config.php`
1. Should it work like that?
2. The new configuration is getting appended at the end of the `wp-config.php` instead of before `/* That's all, stop editing! Happy publishing. */` line. Therefore; the configuration is not taking in consideration **I know this is not in the scope of the thread**

**Comment by juan131** [comment]
> Should it work like that?

Yes, since the **wp-config.php** persisted. So your don't need to append the custom section every time you restart the container.

> The new configuration is getting appended at the end of the **wp-config.php**

Yes, that's expected too. Please correct me if I'm wrong but the line `/* That's all, stop editing! Happy publishing. */` it's just a comment, isn't it? I mean, the new block should also be considered.

**Comment by leandropapasidero** [comment]
As you can see in the following post suggested, you should never declare those constants after the comment.


https://kinsta.com/blog/wp-config-php/

https://help.dreamhost.com/hc/en-us/articles/214693268-WordPress-wp-config-php-overview

I believe the issue will occur since you are declaring the constants after

```
require_once(ABSPATH . 'wp-settings.php');
```

Which set WordPress variables and include files. Therefore those constants will not be available at that time.

**Comment by juan131** [comment]
Hi @leandropapasidero 

That's weird. If you check the current `wp-config.php` sample available at WordPress repository, they add content after `/* That's all, stop editing! Happy publishing. */`:

- https://github.com/WordPress/WordPress/blob/master/wp-config-sample.php

Anyway, I agree that appending the custom section at the end of the configuration file might not be right solution for some scenarios, and we should try to be consistent with that comment.

**Comment by malagant** [comment]
@juan131 What about defining placeholders instead of appending to `wp-config.php`? Especially when defining SSL stuff appending is the wrong approach. I have not much knowledge of Wordpress, but need to set it up for my customers. ;)

**Comment by FraPazGal** [comment]
Hello @malagant,

Thanks for the input! Could you expand on your solution? If I understood it correctly, you are suggesting to add the extra configuration through placeholders defined in `wp-config.php`, is that right? That would indeed solve the problem of having the custom section at the end of the file.

**Comment by adnoh** [comment]
And why not using the tools given like:
`sed -i "/.*That's all, stop editing! Happy publishing.*/i $WORDPRESS_CONFIG_EXTRA" /path/to/wp-config.php`
To insert the content of the env var before the "magic line" ?

**Comment by malagant** [comment]
@adnoh Yes. Also an option for experienced users. I tried ist but then I needed to open #4568 because of missing pieces

**Comment by adnoh** [comment]
@malagant I mixed things up. I knew there is WORDPRESS_EXTRA_WP_CONFIG_CONTENT - so I was thinking of using this and change the code where (I don't know) it is appended to wp-config.php with a sed command like I provided (with WORDPRESS_EXTRA_WP_CONFIG_CONTENT instead of WORDPRESS_CONFIG_EXTRA).
But for a testrive (untested) I would change:
/.nami/components/com.bitnami.wordpress/main.js
Line 95:
`- $file.append(confFile, $app.extraWPConfigContent, {atNewLine: true});`
`+ $file.substitute(confFile, /\/\* That's all, stop editing! Happy publishing. \*\//, $app.extraWPConfigContent + "\n/* That's all, stop editing! Happy publishing. */", {abortOnUnmatch: true});`

I would say the appending to the config needs to get fixed - after the wordpress magic comment line you should not put anything.
There is also more appending happening in main.js which should be refactored in my opinion.

**Comment by FraPazGal** [comment]
Hi @adnoh,

Thanks for sharing your advances! That could indeed be a possible solution, though it would be dependent on that comment not changing. I will update the internal task related to this issue to help the team once we start working on this!

Would that work for you @malagant? You mentioned appending as being an incorrect approach when dealing with SSL, could you expand on that?

**Comment by jfeliweb** [comment]
Hi @FraPazGal,

Any update on this? I am running into an issue with adding config variables that needs to be placed above the "That's all, stop editing! Happy publishing." 

I am open to a solid work around if there's one available.

**Comment by javsalgar** [comment]
Hi,

You could take advantage of the post-init.sh script, which would load a set of scripts to be executed after initialization. With these, you could add as many modifications as you want

https://github.com/bitnami/bitnami-docker-wordpress/blob/master/5/debian-10/rootfs/post-init.sh

Would it work for your use case?

**Comment by marcosbc** [comment]
Hi @jfeliweb, we have just released tag [5.7.1-debian-10-r21](https://github.com/bitnami/bitnami-docker-wordpress#571-debian-10-r21) of the container images which implements a lot of new features. In particular it adds configuration specified via `WORDPRESS_EXTRA_WP_CONFIG_CONTENT` right before the "That's all, stop editing!" comment.
